### PR TITLE
[http] API changes to split path header normalization and forwarding

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

Commit Message: Configuration for path normalization so that normalization can be configured for internal only use and/or forwarding.
Additional Description: Just API changes. `normalize_path` and `merge_slashes` to be deprecated.

There will be normalizations Envoy applies to an incoming path header (`forwarding_transformation`) that "permanently" transform an incoming path header and affect the forwarded path header. Internal transformations (`filter_transformations`) apply only within Envoy processing. If an intermediate filter changes the internal path header, we don't restore the header.

image

The plan:
* Make `ConnManagerUtility` and `Router` friend classes with `HeaderMap`
* Create private `HeaderMap` methods: `setNormalizedPath` that `ConnManagerUtility` calls prior to filter processing and routing that sets private members `original_path_` and `normalized_path_`. Then `restoreOriginalPathIfUnchanged` called by `Router` if we are doing internal-only normalization that restores `:path` with the original header only filters didn't change the path.
* No HTTP filters have access to the original path for forwarding to prevent misuse. 

Risk Level: Low, just API
Release notes: TODO on implementation
Runtime Guard: TODO on implementation. Runtime guard to overwrite forwarding `http_connection_manager.forward_normalized_path` (can revert right now)
Starts: https://github.com/envoyproxy/envoy/issues/6589

### CODE REVIEW

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 1]
```diff
@@ -238,6 +238,52 @@ message HttpConnectionManager {
...
+  message PathNormalizationOptions {
+    // Defines a policy to apply normalization.  Path rewrites :ref:`regex_rewrite
+    // <envoy_api_field_route.RouteAction.regex_rewrite>` or :ref:`prefix_rewrite
+    // <envoy_api_field_route.RouteAction.prefix_rewrite>` will apply to the
+    // *:path* header destined for the upstream.
+    enum NormalizationPolicy {
+      // Normalizations apply internally before any processing of requests by HTTP
+      // filters, routing, and matching, *and* will affect the upstream *:path* header
+      // as well.
+      FORWARD = 0;
```
Comment (adisuissa) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:255]: Should `FORWARD` be the default behavior?
I'm not sure what the intended default should be, and in this case it might make more sense to add a new value `POLICY_UNSPECIFIED = 0` if the policy should be explicit, and increase the values of `FORWARD` and `INTERNAL` by 1.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 2]
```diff
@@ -238,6 +238,52 @@ message HttpConnectionManager {
...
+    // See `Normalization and Comparison
+    // <https://tools.ietf.org/html/rfc3986#section-6>`_ for details of
+    // normalization.  Note that Envoy does not perform `case normalization
+    // <https://tools.ietf.org/html/rfc3986#section-6.2.2.1>`_
+    google.protobuf.BoolValue normalize_path = 2;
+
+    // Determines if adjacent slashes in the path are merged into one before any
+    // processing of requests by HTTP filters or routing. Without setting this
+    // option, incoming requests with path `//dir///file` will not match against
+    // route with `prefix` match set to `/dir`. Defaults to `false`. Note that
```
Comment (adisuissa) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:281]: According to the [API_STYLE guide](https://github.com/envoyproxy/envoy/blob/1d1b708c7bf6efa02c41d9ce22cbf1e4a1aeec2c/api/STYLE.md#api-style-guidelines), the wrapped scalar type should be used in cases that the default value isn't compatible with proto3 default (false in this case). Is there an explicit reason for using the wrapper rather than `bool`?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 1 above)
Comment (adisuissa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:255]: Just saw in the doc update that `FORWARD` is the intended default behavior.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 3]
```diff
@@ -581,6 +627,12 @@ message HttpConnectionManager {
   // *not* the deprecated but similarly named :ref:`stream_error_on_invalid_http_messaging
   // <envoy_v3_api_field_config.core.v3.Http2ProtocolOptions.stream_error_on_invalid_http_messaging>`
   google.protobuf.BoolValue stream_error_on_invalid_http_message = 40;
+
+  // [#not-implemented-hide:] Path normalization configuration. This includes
+  // configurations for transformations (e.g. RFC 3986 normalization or merge
+  // adjacent slashes) and the policy to apply them. The policy determines
+  // whether transformations affect the forwarded *:path* header.
+  PathNormalizationOptions path_normalization_options = 43;
```
Comment (adisuissa) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:635]: Should there be a default policy if this field isn't part of a configuration?

**Review by adisuissa — CHANGES_REQUESTED** [review]
Thanks for working on this!

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 4]
```diff
@@ -238,6 +238,52 @@ message HttpConnectionManager {
     google.protobuf.BoolValue enabled = 3;
   }
 
+  // Transformations that apply to path headers. The policy to apply these
+  // transformations can be tuned by the configuration field. Transformations
+  // can be applied internally only, so that the normalized header is used in
+  // matching, routing, and HTTP filters while the original path header is
+  // forwarded upstream, or additionally apply to the forwarded *:path* header.
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:245]: The last disjunct in this sentence is a bit hard to parse. Suggest explaining the forwarding distinct in a following sentence.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 5]
```diff
@@ -238,6 +238,52 @@ message HttpConnectionManager {
     google.protobuf.BoolValue enabled = 3;
   }
 
+  // Transformations that apply to path headers. The policy to apply these
+  // transformations can be tuned by the configuration field. Transformations
+  // can be applied internally only, so that the normalized header is used in
+  // matching, routing, and HTTP filters while the original path header is
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:244]: What about observability (logging, tracing)?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 6]
```diff
@@ -238,6 +238,52 @@ message HttpConnectionManager {
...
+  // forwarded upstream, or additionally apply to the forwarded *:path* header.
+  message PathNormalizationOptions {
+    // Defines a policy to apply normalization.  Path rewrites :ref:`regex_rewrite
+    // <envoy_api_field_route.RouteAction.regex_rewrite>` or :ref:`prefix_rewrite
+    // <envoy_api_field_route.RouteAction.prefix_rewrite>` will apply to the
+    // *:path* header destined for the upstream.
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:250]: Why only mention these two?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 7]
```diff
@@ -238,6 +238,52 @@ message HttpConnectionManager {
...
+      // Normalizations apply internally before any processing of requests by HTTP
+      // filters, routing, and matching, *and* will affect the upstream *:path* header
+      // as well.
+      FORWARD = 0;
+
+      // Normalizations only apply internally. The original *:path* header will
+      // be sent upstream *unless* the path has been mutated by a filter. Any
+      // regex rewrites that the router performs will be on the upstream *:path*
+      // header.
+      INTERNAL = 1;
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:261]: I wonder if there is a case for being able to orthogonally specify both the match and forward behavior? E.g. someone wants to be able to do RFC3986 on both but only merge_slashes on match? It seems a bit contrived, but as we add more normalizations, maybe it becomes interesting. I think code/API wise it's not hard to support this dual config.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:281]: This allows us to change defaults in future..

**Review by htuch — COMMENTED** [review]
Awesome, thanks for tackling this important feature!

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 7 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:261]: Yes, I did consider this, especially with the mindset of making this as general as possible. It comes with the downside of having to specify the behavior for each transformation. I wonder if users can either specify a catch all policy OR a policy for each transformation if they want a very customizeable experience. Something like:

PathNormalizationOptions:
    - merge_slashes: true
      policy: internal
    - path_normalization: true
      policy: forward

PathNormalizationOptions:
    - uniform_policy: forward
    - merge_slashes: true
    - path_normalization: true

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 6 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:250]: Those are the only router transformations, but that may change in the future so I will change the language.

I mentioned that in this comment, it's only relevant for `INTERNAL`:
> // Normalizations only apply internally. The original *:path* header will
      // be sent upstream *unless* the path has been mutated by a filter.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 3 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:662]: Specified! (RFC 3986 by default, policy forward by default)

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 5 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: It depends on the context. Only the normalized path is visible internally so if a filter logs headers, they will see the normalized header. If we finalize spans before forwarding to an upstream, we see the internal header, otherwise we see the forwarded header. This can obviously be changed in the code.

I'm not really sure what the right answer here is. If it should always be the "forwarded header" or always the original header, then we lose out on observability in some context. I think we either have to keep the logging context dependent as is, OR we allow `HttpTracerUtility` visibility into the original header (and for logs, I don't know -- as soon as we open the original header to logs, we open it up to filters).

e.g. if you log the request headers directly prior to forwarding, you will see the to-be-forwarded header, like here https://github.com/envoyproxy/envoy/blob/280654be74c4760a7f61f3658b035e4800f2deca/source/common/router/router.cc#L583. before that, you see the internal header.

**Review by adisuissa — DISMISSED** [review]
LGTM! Thanks for working on this!

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 8]
```diff
@@ -238,6 +238,75 @@ message HttpConnectionManager {
...
+    }
+
+    // Should paths be normalized according to RFC 3986 before any processing of
+    // requests by HTTP filters or routing? For paths that fail this check,
+    // Envoy will respond with 400 to paths that are malformed. See
+    // `Normalization and Comparison
+    // <https://tools.ietf.org/html/rfc3986#section-6>`_ for details of
+    // normalization. Note that Envoy does not perform `case normalization
+    // <https://tools.ietf.org/html/rfc3986#section-6.2.2.1>`_
+    message NormalizePath {
```
Comment (asraa) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:271]: @adisuissa Do you think I should combine these messages?
e.g.
```
message NormalizationControl {
    NormalizationPolicy policy = 1;
    google.protobuf.BoolValue value = 2;
}
// RFC 3986
NormalizationControl normalize_path = 3;
// Adjacent slashes
NormalizationControl merge_slashes = 3;
```

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 9]
```diff
@@ -238,6 +238,75 @@ message HttpConnectionManager {
...
+    }
+
+    // Wholesale policy that applies to each option if they do not specify
+    // individual policies. Individual policies allow configuring different
+    // behaviors for each transformation (e.g. forward RFC 3986 path
+    // normalization, use merge_slashes for matching only).
+    // Defaults to FORWARD.
+    NormalizationPolicy all_policy = 1;
+
+    // Message to configure RFC 3986 path normalization. Be fault, this is
```
Comment (adisuissa) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:300]: Nit: typo: "fault" should be "default".

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 8 above)
Comment (adisuissa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:271]: I agree that both can be merged into `NormalizationControl` type, but IIUC the default value for the `BoolValue value` needs to be different for path-normalization and for slash-merging, no?
Also, if you think that in the future, path normalization will have sub-policy settings (e.g., disabling percent encoding), then it's better to keep them separate, and add a comment.

Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:271]: Defaults could always be tuned with the param in `PROTOBUF_GET_WRAPPED_OR_DEFAULT`
Hmm after thinking about it, maybe some need individual controls, but a vast majority won't. I suppose `normalize_path` may need sub-options (for windows path?), but that's the only example I can think of.

Comment (adisuissa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:271]: Another option is to add a new enum value: `DISABLE` to `NormalizationPolicy`. IIUC, this keeps the same semantics, removes the BoolValue flag, and potentially makes it easier to configure.
But if additional sub-options are needed then specific messages also make sense.

Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:271]: That is a great idea! Let me try that out, I'll push it if it looks good. 
My feeling is that sub-options can always be made in to separate options (see waf transformation knobs https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-(v2.x)#Transformation_functions)

Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: I think I need to wrap enums with a WKT though.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 10]
```diff
@@ -238,6 +238,62 @@ message HttpConnectionManager {
...
+  message PathNormalizationOptions {
+    // Defines a policy to apply to a normalization. Any path rewrites that the
+    // router performs (e.g. :ref:`regex_rewrite
+    // <envoy_api_field_config.route.v3.RouteAction.regex_rewrite>` or
+    // :ref:`prefix_rewrite
+    // <envoy_api_field_config.route.v3.RouteAction.prefix_rewrite>`) will apply
+    // to the *:path* header destined for the upstream.
+    enum NormalizationPolicy {
+      // Unspecified. See individual transformations for defaults.
+      UNSPECIFIED = 0;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:260]: Maybe this has been already discussed but I would probably call this DEFAULT?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 11]
```diff
@@ -238,6 +238,62 @@ message HttpConnectionManager {
...
+  // for encoding to the upstream (e.g. when closing the stream after a response
+  // is received from an upstream) will show the forwarded header.
+  message PathNormalizationOptions {
+    // Defines a policy to apply to a normalization. Any path rewrites that the
+    // router performs (e.g. :ref:`regex_rewrite
+    // <envoy_api_field_config.route.v3.RouteAction.regex_rewrite>` or
+    // :ref:`prefix_rewrite
+    // <envoy_api_field_config.route.v3.RouteAction.prefix_rewrite>`) will apply
+    // to the *:path* header destined for the upstream.
+    enum NormalizationPolicy {
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:258]: Do we ever envision having params for any of this? I'm worried that enum here might be fragile. Should we instead do a oneof with a message for each option, even if that message is empty for now? This would be easier to future proof IMO?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 12]
```diff
@@ -238,6 +238,62 @@ message HttpConnectionManager {
...
+    // <https://tools.ietf.org/html/rfc3986#section-6.2.2.1>`_
+    NormalizationPolicy normalize_path = 1;
+
+    // Determines if adjacent slashes in the path are merged into one before any
+    // processing of requests by HTTP filters or routing. Defaults to
+    // DISABLE. Without setting this option, incoming requests with path
+    // `//dir///file` will not match against route with `prefix` match set to
+    // `/dir`. Note that slash merging is not part of `HTTP spec
+    // <https://tools.ietf.org/html/rfc3986>`_ and is provided for convenience.
+    NormalizationPolicy merge_slashes = 2;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:294]: This might be getting into the implementation weeds, but do you envision any issues if someone selects INTERNAL for one and FORWARD for another? Is it going to lead to complicated state tracking?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks just a few API questions.

/wait

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 12 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:294]: It's open to do so.

If the `HeaderMap` privately stores the original path, then setting the FORWARD header could be independent of the internal transformed path. (However, we do have to store the original internal header too, to make sure we don't override the path if a filter changed it)

`HeaderMap`:
   * Path: public visible path, HCM sets to INTERNAL path, router sets to FORWARD when forwarding
   * private: `original_path` used to create FORWARD path
   * private: `internal_path` the internal path set by the HCM. used to compare the current path in the router overriding HeaderMap::Path() with FORWARD header.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 11 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:258]: Yeah, @adisuissa brought that up. I'll change to message per option (empty if no params). 

By the way, @yangminzhu has been creating a proposal to add these types of transformations (ModSecurity based https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-(v2.x)#Transformation_functions) to `StringMatcher` as well so that matching can be done in filters with added transformations (e.g. we already have the `lowercase` transformation with `ignore_case` in StringMatcher).

I was planning to change this after talking with @yangminzhu to define a core message `TextTransformations` with repeated operations (a oneof with message per option), and consuming that here for a list of internal transformations and forwarding transformations?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 12 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: I added a comment about `internal_transformations`:

```
    // These will be applied after :ref:`full_transformation
    // <envoy_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.PathNormalizationOptions.full_transformation`
    // is applied. The *:path* header before this transformation will be restored in the router
    // filter and sent upstream unless it was mutated by a filter```

File: api/envoy/type/v3/transformations.proto [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  message Operation {
+    // No operation.
+    message None {
+    }
+
+    // Should paths be normalized according to RFC 3986 before any processing of requests by HTTP
+    // filters or routing? See `Normalization and Comparison
+    // <https://tools.ietf.org/html/rfc3986#section-6>`_ for details of normalization. Note that Envoy
+    // does not perform `case normalization <https://tools.ietf.org/html/rfc3986#section-6.2.2.1>`_
+    message NormalizePath {
```
Comment (yangminzhu) [inline api/envoy/type/v3/transformations.proto:34]: I understand this links to the RFC but it would be super useful to actually also document the exact normalization behavior in Envoy, the RFC is really long, has a lot cross-references and sometimes a bit vague regrading the exact normalization behavior.

Also should this be renamed to `NormalizePathPerRFC3986`?

File: api/envoy/type/v3/transformations.proto [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    // Determines if adjacent slashes are merged into one Without setting this option, incoming
```
Comment (yangminzhu) [inline api/envoy/type/v3/transformations.proto:37]: This is a general text transformation API, I think the comment should also be more generic, it could mention the most common use case is for request path, but that is just one use case and this transformation could be used against any input. Similar to the `NormalizaPath` above.

File: api/envoy/type/v3/transformations.proto (hunk 13 above)
Comment (asraa) [reply api/envoy/type/v3/transformations.proto:34]: > Also should this be renamed to NormalizePathPerRFC3986?

Is this in anticipation of other path normalizations? If so, should those be extended inside the message?

Comment (yangminzhu) [reply api/envoy/type/v3/transformations.proto:34]: yes, in case we have other path normalization but I'm not sure what are they for now. I'm not sure if it will create some difficulties when later there are more fields inside the message, like the order and relationship among them?

Comment (asraa) [reply api/envoy/type/v3/transformations.proto:34]: Yeah, I agree separate options are better.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 12 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: @chaoqin-li1123 suggested that we should clarify this too. `full_transformation` (previously called FORWARD) should be though of as incoming header normalization that mutates the header before processing and affects the forwarded.

internal ones only exist temporarily in Envoy during processing. we forward the path with the incoming header normalization.

I am happy to make the name change from `full_transformation` to `incoming_transformation` if that's clearer.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 15]
```diff
@@ -238,6 +239,33 @@ message HttpConnectionManager {
...
 
+  // [#not-implemented-hide] Transformations that apply to path headers. Transformations are applied
+  // before any processing of requests by HTTP filters, routing, and matching. Only the normalized
+  // path will be visible internally if a transformation is enabled. Any path rewrites that the
+  // router performs (e.g. :ref:`regex_rewrite
+  // <envoy_api_field_config.route.v3.RouteAction.regex_rewrite>` or :ref:`prefix_rewrite
+  // <envoy_api_field_config.route.v3.RouteAction.prefix_rewrite>`) will apply to the *:path* header
+  // destined for the upstream.
+  //
+  // Note: access logging and tracing will show the original *:path* header.
```
Comment (yanavlasov) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:250]: I wonder if we also need to add a way to log normalized path header. This is outside of the scope of this PR, but maybe add an Issue to track this.

**Review by yanavlasov — COMMENTED** [review]
Awesome changes.
/wait

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 15 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:250]: Yes, I imagine we can add a formatter `%UPSTREAM_PATH%` so users can see the path forwarded, and something to log the internal path used for matching. @chaoqin-li1123 FYI.

I added a comment here https://github.com/envoyproxy/envoy/issues/6589#issuecomment-786665159
when implementation starts we should file a followup

**Comment by asraa** [comment]
@mattklein123 could you please take another look? Controls have individual messages, hopefully the internal/full transformation interaction is more clear.

File: api/envoy/type/v3/transformations.proto [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+// TextTransformations define an API to apply a sequence of operations that can
+// be used to alter text before it is used for matching or routing. Multiple
+// actions can be applied in the same Transformation, forming a sequential
+// pipeline. The transformations will be performed in the order that they
+// appear.
+//
+// This API is a work in progress.
+
+message TextTransformation {
```
Comment (mattklein123) [inline api/envoy/type/v3/transformations.proto:23]: Is this message name a bit too generic? Should perhaps this be in some HTTP specific namespace and named more specifically for HTTP normalization?

File: api/envoy/type/v3/transformations.proto [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    // <envoy_api_msg_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.PathNormalizationOptions>`
+    // will allow incoming requests with path `//dir///file` to match against route with `prefix`
+    // match set to `/dir`. When using for header transformations, note that slash merging is not
+    // part of `HTTP spec <https://tools.ietf.org/html/rfc3986>`_ and is provided for convenience.
+    message MergeSlashes {
+    }
+
+    oneof Kind {
+      // No transformation.
+      None none = 1;
```
Comment (mattklein123) [inline api/envoy/type/v3/transformations.proto:51]: Do we need the None type? I think if you make the oneof required we can just require one of the known types otherwise the user just wouldn't add an operation?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks at a high level this LGTM with some small remaining comments. 

/wait

File: api/envoy/type/v3/transformations.proto (hunk 16 above)
Comment (asraa) [reply api/envoy/type/v3/transformations.proto:23]: @yangminzhu What do you think? StringMatcher should be able to consume it to normalize/transform on matching. These are mostly, but not completely, header values.

Comment (yangminzhu) [reply api/envoy/type/v3/transformations.proto:23]: I proposed the name `TextTransformation` for the purpose that it can be used as a generic text transformation API, which could be used for things more than just HTTP normalization (e.g. remove prefix, base64Decode, etc.) and embed directly in StringMatcher, etc.

Comment (mattklein123) [reply api/envoy/type/v3/transformations.proto:23]: I don't feel very strongly about it either way. @envoyproxy/api-shepherds any opinions?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 18]
```diff
@@ -238,6 +239,33 @@ message HttpConnectionManager {
...
+  message PathNormalizationOptions {
+    // [#not-implemented-hide] Normalization applies internally before any processing of requests by
+    // HTTP filters, routing, and matching *and* will affect the forwarded *:path* header. Defaults
+    // to :ref:`NormalizePathRFC3986
+    // <envoy_api_msg_type.v3.TextTransformation.Operation.NormalizePathRFC3986>`. When not specified,
+    // this value may be overridden by the runtime variable
+    // :ref:`http_connection_manager.normalize_path<config_http_conn_man_runtime_normalize_path>`. For
+    // paths that fail this check, Envoy will respond with 400 to paths that are malformed.
+    type.v3.TextTransformation full_transformation = 1;
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:259]: Nit: I think `full_transformation` is a bit vague. I suggest renaming to `forwarding_transformation`.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 19]
```diff
@@ -238,6 +239,33 @@ message HttpConnectionManager {
...
+
+    // [#not-implemented-hide] Normalization only applies internally before any processing of
+    // requests by HTTP filters, routing, and matching. These will be applied after full
+    // transformation is applied. The *:path* header before this transformation will be restored in
+    // the router filter and sent upstream unless it was mutated by a filter. Defaults to no
+    // transformations.
+    type.v3.TextTransformation internal_transformation = 2;
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:266]: Nit: prefer `match_transformation` or `http_filter_transformation` over internal..

File: api/envoy/type/v3/transformations.proto (hunk 16 above)
Comment (htuch) [reply api/envoy/type/v3/transformations.proto:23]: Why are we putting HTTP specific things in a generic `TextTransformation`? I think just having an HTTP specific `PathTransformation` would make sense, and distinct messages for other things (e.g. normalizing headers, etc.)

File: api/envoy/type/v3/transformations.proto [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+    // Determines if adjacent slashes are merged into one. A common use case is for a request path
+    // header. Using this option in `:ref: PathNormalizationOptions
+    // <envoy_api_msg_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.PathNormalizationOptions>`
+    // will allow incoming requests with path `//dir///file` to match against route with `prefix`
+    // match set to `/dir`. When using for header transformations, note that slash merging is not
+    // part of `HTTP spec <https://tools.ietf.org/html/rfc3986>`_ and is provided for convenience.
+    message MergeSlashes {
+    }
+
+    oneof Kind {
```
Comment (htuch) [inline api/envoy/type/v3/transformations.proto:45]: Nit: prefer `operation_specifier` or something (don't capitalize).

File: api/envoy/type/v3/transformations.proto [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+      // Enable path normalization per RFC 3986.
+      NormalizePathRFC3986 normalize_path_rfc_3986 = 2;
+
+      // Enable merging adjacent slashes.
+      MergeSlashes merge_slashes = 3;
+    }
+  }
+
+  // A list of operations to apply. Transformations will be performed in the order that they appear.
```
Comment (htuch) [inline api/envoy/type/v3/transformations.proto:56]: Is this expressive enough to capture notions of "keep escape encoding until we reach a fixed point?". Would such a transform be interesting?

**Review by htuch — COMMENTED** [review]
Looks great. A few bike shed comments.

File: api/envoy/type/v3/transformations.proto (hunk 16 above)
Comment (asraa) [reply api/envoy/type/v3/transformations.proto:23]: I'm going to make it distinct for each. I will unwrap it one layer so a `PathTransformation` is oneof a path-relevant operation kind. I can consume `repeated PathTransformation` in HCM, and you can consume a `PathTransformation` as one of a `TextTransformation` in `StringMatcher`. We can also add mutual types to our messages from a common place.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 18 above)
Comment (asraa) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:259]: Sure, my worry was that it will be misinterpreted as a normalization that occurs on forwarding (and not before).

File: api/envoy/type/v3/transformations.proto (hunk 21 above)
Comment (asraa) [reply api/envoy/type/v3/transformations.proto:56]: Double-decoding is definitely interesting. Triple-decoding is less likely, I think a bypass would have to mean that there may be some decoding hops on the backend but I have seen this happen.

In terms of API, a decode message could have a notion of repetition. I don't think it should be a top-level option (`urlDecode: - double`)

File: api/envoy/type/v3/transformations.proto (hunk 16 above)
Comment (yangminzhu) [reply api/envoy/type/v3/transformations.proto]: sorry to re-open the discussion here, @htuch @asraa 

I think `TextTransformation` is a general class that can be used in many places, here at HTTP is just one use case, and path is just one use case, it could be host, HTTP header, some claim in JWT token or even a field in the dynamic metadata.

Not every transformation is needed for every of these cases, it's up to the user to choose which one they want to use.

I'm just a bit worried that making it `PathTransformation` is too restrictive and makes it hard to be reused in all other cases (e.g. host, header and etc.)

File: api/envoy/type/v3/path_transformations.proto [hunk 22]
```diff
@@ -0,0 +1,57 @@
+syntax = "proto3";
+
+package envoy.type.v3;
```
Comment (mattklein123) [inline api/envoy/type/v3/path_transformations.proto:3]: Final nit from me: maybe move this into an HTTP namespace? Perhaps envoy.type.http.v3 or something? Whatever you think is best.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with one small comment.

/wait

File: api/envoy/type/v3/transformations.proto (hunk 21 above)
Comment (yangminzhu) [reply api/envoy/type/v3/transformations.proto]: I agree, detail transformation can have its detail configuration inside its message, so a urlDecode transformation can specify such configuration internally.

File: api/envoy/type/v3/transformations.proto (hunk 16 above)
Comment (htuch) [reply api/envoy/type/v3/transformations.proto]: We can add other transformations as well for these things. We don't want to reuse RFC3986 or merge slashes. I think having specialized messages for these distinct normalization is a good thing, e.g. `PathTransformation`, `HeaderKeyTransformation`, `HeaderValueTransformation`. Agree with @mattklein123 on namespace.

LGTM modulo Matt's comment ,will hand it over to @mattklein123 for merge.

**Review by mattklein123 — APPROVED** [review]
Thanks!
